lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -897,59 +901,106 @@ _evaluateMinimumFee
     :: Cardano.ProtocolParameters
     -> SealedTx
     -> Maybe Coin
-_evaluateMinimumFee pp tx =
-    fromCardanoLovelace <$> minFee
+_evaluateMinimumFee pp tx = do
+    anyShelleyTx <- inAnyShelleyBasedEra (cardanoTx tx)
+    pure $ withShelleyBasedBody anyShelleyTx $ \_era body ->
+        fromCardanoLovelace $
+            Cardano.evaluateTransactionFee
+                pp
+                body
+                (estimateNumberOfWitnesses body)
+                0
+
+-- | Estimate the size of the transaction (body) when fully signed.
+_estimateSignedTransactionSize
+    :: Cardano.IsShelleyBasedEra era
+    => Cardano.ProtocolParameters
+    -> Cardano.TxBody era
+    -> TxSize
+_estimateSignedTransactionSize pparams body =
+    let
+        nWits = estimateNumberOfWitnesses body
+
+        -- Hack which allows us to rely on the ledger to calculate the size of
+        -- witnesses:
+        feeOfWits = minfee nWits - minfee 0
+        sizeOfWits =
+            case feeOfWits `quotRem` perByte of
+                (n, 0) -> TxSize n
+                (_, _) -> error $ unwords
+                    [ "evaluateTransactionSize:"
+                    , "couldn't divide"
+                    , show feeOfWits
+                    , "lovelace"
+                    , "(the fee contribution"
+                    , "of"
+                    , show nWits
+                    , "witesses),"
+                    , "with"
+                    , show perByte
+                    , "lovelace/byte"
+                    ]
+        sizeOfTx = TxSize
+            . fromIntegral
+            . BS.length
+            . serialisedTx
+            $ sealedTxFromCardanoBody body
+    in
+        sizeOfTx <> sizeOfWits
+  where
+    minfee nWits = Coin.toNatural $ fromCardanoLovelace $
+            Cardano.evaluateTransactionFee
+                pparams
+                body
+                nWits
+                0
+    perByte = view #protocolParamTxFeePerByte pparams
+
+-- | Estimate number of shelley era witnesses
+--
+-- NOTE: Assuming one witness per certificate is wrong. KeyReg certs don't
+-- require witnesses, and several certs may share the same key.
+--
+-- NOTE: Similar to 'estimateTransactionKeyWitnessCount' from cardano-api, which
+-- we cannot use because it requires a 'TxBodyContent BuildTx era'.
+estimateNumberOfWitnesses
```

Comment:
Is this function guaranteed to produce an upper bound, i.e., a value that is greater than equal to the number of witnesses eventually required?

Reply:
Yes (although strictly speaking it is not tested)

I elaborated in 5c7d735f3 and added a task for reducing the over-estimation. The overestimation could actually be significant when used with a single address wallet ((nIns + nCol) witnesses vs 1 witness)